src/main.c

```diff
@@ -1222,9 +1222,56 @@ static int on_config_max_connections(h2o_configurator_command_t *cmd, h2o_config
 
 static int on_config_num_threads(h2o_configurator_command_t *cmd, h2o_configurator_context_t *ctx, yoml_t *node)
 {
-    if (h2o_configurator_scanf(cmd, node, "%zu", &conf.num_threads) != 0)
+    size_t num_threads = 0;
+    free(conf.thread_map.entries);
+    memset(&conf.thread_map, 0, sizeof(conf.thread_map));
+    if (node->type == YOML_TYPE_SCALAR) {
+        int i;
+        if (h2o_configurator_scanf(cmd, node, "%zu", &num_threads) != 0)
+            return -1;
+        h2o_vector_reserve(NULL, &conf.thread_map, num_threads);
+        for (i = 0; i < num_threads; i++)
+            conf.thread_map.entries[conf.thread_map.size++] = -1;
+    } else if (node->type == YOML_TYPE_SEQUENCE) {
+#ifndef H2O_HAS_PTHREAD_SETAFFINITY_NP
+        fprintf(stderr, "[error] Can't handle a CPU list, this platform doesn't support `pthread_setaffinity_np`\n");
         return -1;
-    if (conf.num_threads == 0) {
+#endif
+        /* a sequence is treated as a list of CPUs to bind to, one per thread to instantiate */
+        int i;
+        for (i = 0; i < node->data.sequence.size; i++) {
+            char *cpu_spec;
+            if (node->data.sequence.elements[i]->type != YOML_TYPE_SCALAR) {
+                h2o_configurator_errprintf(cmd, node, "CPUs in cpu sequence must be a scalar");
+                return -1;
+            }
+            cpu_spec = node->data.sequence.elements[i]->data.scalar;
+            if (index(cpu_spec, '-') == NULL) {
+                int pos, cpu_num;
+                if (sscanf(cpu_spec, "%d%n", &cpu_num, &pos) != 1 || pos != strlen(cpu_spec)) {
+                    h2o_configurator_errprintf(cmd, node, "Invalid CPU number: CPUs must be specified as a positive number or as a range of positive numbers");
+                    return -1;
+                }
+                h2o_vector_reserve(NULL, &conf.thread_map, conf.thread_map.size + 1);
+                conf.thread_map.entries[conf.thread_map.size++] = cpu_num;
+            } else {
+                int pos, cpu_low, cpu_high, cpu_num;
+                if (sscanf(cpu_spec, "%d-%d%n", &cpu_low, &cpu_high, &pos) != 2 || pos != strlen(cpu_spec) || cpu_low < 0 || cpu_high < 0) {
```

Comment:
Would it make sense to use `%u` instead? We would not want to see something like `-3-1` being accepted.